src/full/Agda/TypeChecking/Conversion.hs

```diff
@@ -319,24 +324,26 @@ compareTerm' cmp a m n =
               mNeutral <- isNeutral m
               n <- reduceB n
               nNeutral <- isNeutral n
-              case (m, n) of
-                _ | isMeta m || isMeta n ->
-                    compareAtom cmp (AsTermsOf a') (ignoreBlocking m) (ignoreBlocking n)
-
-                _ | mNeutral && nNeutral -> do
-                    -- Andreas 2011-03-23: (fixing issue 396)
-                    -- if we are dealing with a singleton record,
-                    -- we can succeed immediately
-                    ifM (isSingletonRecordModuloRelevance r ps) (return ()) $
-                      -- do not eta-expand if comparing two neutrals
-                      compareAtom cmp (AsTermsOf a') (ignoreBlocking m) (ignoreBlocking n)
-                _ -> do
-                  (tel, m') <- etaExpandRecord r ps $ ignoreBlocking m
-                  (_  , n') <- etaExpandRecord r ps $ ignoreBlocking n
-                  -- No subtyping on record terms
-                  c <- getRecordConstructor r
-                  -- Record constructors are covariant (see test/succeed/CovariantConstructors).
-                  compareArgs (repeat $ polFromCmp cmp) [] (telePi_ tel __DUMMY_TYPE__) (Con c ConOSystem []) m' n'
+              isSingleton <- isSingletonRecordModuloRelevance r ps
```

Comment:
This is an expensive operation that shouldn't be done needlessly.  So it should not be pulled to the toplevel when there is a possibility that it is not used subsequently (meta/meta case, e.g.).

Also, it seems that changing the logic leads to new unsolved metas (see CI).  E.g. if one side is a meta (but shortcut failed), and now you succeed immediately due to singleton record, you might miss the solution of the meta.

Reply:
(Also, a change in the constraint solver should be a separate PR, if at all.  But I guess the semantics change was not intended.)